Adding support for `RecoverKey` in the chain interface.

chatton opened the pull request:
## Description

This PR adds the functionality to recover a test user from a given mnemonic.

## Checklist

- [X] I have made sure the upstream branch for this PR is correct
- [X] I have made sure this PR is ready to merge
- [x] I have made sure that I have assigned reviewers related to this project

## Changes

- [x] Added `RecoverKey` to `ChainNode` and `CosmosChain` types.
- [x] Added `RecoverKey` to `Chain` interface.
- [ ] ~Added `Shell` field to `ChainConfig`. This is to allow you to specify which shell commands which run using `-c <script>` will use, as not all images will have the same shells.~
- [x] Added `GetAndFundTestUserWithMnemonic` which generates a user using a given mnemonic.
- [x] Changed `GetAndFundTestUsers` to delegate to `GetAndFundTestUserWithMnemonic`
- [x] Changed `generateUserWallet` to either `Create` or `Recover` based on the presence of a mnemonic.


## Testing

Tests have been added which verify that users can be created using `CreateKey` and `RecoverKey`


closes #169

chatton commented on chain/penumbra/penumbra_chain.go:
```diff
@@ -140,6 +140,10 @@ func (c *PenumbraChain) CreateKey(ctx context.Context, keyName string) error {
 	return c.getRelayerNode().PenumbraAppNode.CreateKey(ctx, keyName)
 }
 
+func (c *PenumbraChain) RecoverKey(ctx context.Context, name, mnemonic string) error {
+	return fmt.Errorf("RecoverKey not implemented for PenumbraChain")
+}
```
I'm not sure exactly what would need to be done here.

mark-rushakoff replied:
I think this error is fine for now.

mark-rushakoff commented on test_user.go:
```diff
@@ -37,6 +44,31 @@ func generateUserWallet(ctx context.Context, keyName string, chain ibc.Chain) (*
+		Denom:   chainCfg.Denom,
+	})
+	require.NoError(t, err, "failed to get funds from faucet")
+	require.NoError(t, test.WaitForBlocks(ctx, 5, chain), "failed to wait for blocks")
```
Did 5 come from anywhere special? Would 4 or 3 work?

Some of these tests can take a long time to complete, so if there is a way to minimize time spent waiting, we should take it.

Maybe it would make more sense to not wait inside this function, and indicate in Godoc that the caller should wait for some blocks to complete before the funds will be accessible. There is one wait at the end of `GetAndFundTestUsers` that would be redundant with this call, if it stays.

chatton replied:
This wait was just taken from the existing `GetAndFundTestUsers` function, however I think it's no harm removing it and letting the caller know as you suggested. I can remove from both.

mark-rushakoff commented on interchain_test.go:
```diff
@@ -145,6 +148,83 @@ func TestInterchain_GetRelayerWallets(t *testing.T) {
+		require.NotEmpty(t, users[0].KeyName)
+	})
+
+	_ = ic.Close()
```
Preferable to `defer ic.Close()` immediately after the initial assignment. 

If the linter complains about a discarded error value, `defer func() { _ = ic.Close() }` is fine.

mark-rushakoff commented on ibc/types.go:
```diff
@@ -62,6 +63,14 @@ func (c ChainConfig) MergeChainSpecConfig(other ChainConfig) ChainConfig {
 	return c
 }
 
+// GetShell returns the shell that should be used when executing scripts.
+func (c ChainConfig) GetShell() string {
+	if c.Shell == "" {
+		return "sh"
+	}
+	return c.Shell
+}
+
```
For now, I think we should assume all containers have `sh` available. It will likely be a long time before we run into something we can't do with plain `sh`.

mark-rushakoff commented on interchain_test.go:
```diff
@@ -145,6 +148,83 @@ func TestInterchain_GetRelayerWallets(t *testing.T) {
+		t, pool, network, home,
+	)
+
+	ic := ibctest.NewInterchain().
+		AddChain(gaia0).
+		AddChain(gaia1).
+		AddRelayer(r, "r").
+		AddLink(ibctest.InterchainLink{
+			Chain1:  gaia0,
+			Chain2:  gaia1,
+			Relayer: r,
+		})
```
Because this test is only exercising funding a user, I think this can be reduced to `ic := ibctest.NewInterchain().AddChain(gaia0)`. 

This should be testable even without an `Interchain`, but since that does the heavy lifting to start a chain, it's probably more convenient than calling several methods directly on a chain.

mark-rushakoff commented on interchain_test.go:
```diff
@@ -145,6 +148,83 @@ func TestInterchain_GetRelayerWallets(t *testing.T) {
+		NetworkID: network,
+	}))
+
+	t.Run("Test User creation with mnemonic", func(t *testing.T) {
```
Names of subtests should be shorter because they are displayed after the outer test's name. This line can be `t.Run("with mnemonic", ...)`.

mark-rushakoff commented on interchain_test.go:
```diff
@@ -145,6 +148,83 @@ func TestInterchain_GetRelayerWallets(t *testing.T) {
+		require.NotEmpty(t, user.KeyName)
+	})
+
+	t.Run("Test user creation without mnemonic", func(t *testing.T) {
```
`t.Run("without mnemonic", ...)`

mark-rushakoff commented on interchain_test.go:
```diff
@@ -145,6 +148,83 @@ func TestInterchain_GetRelayerWallets(t *testing.T) {
+		user := ibctest.GetAndFundTestUserWithMnemonic(t, ctx, keyName, mnemonic, 10000, gaia0)
+
+		require.NotEmpty(t, user.Address)
+		require.NotEmpty(t, user.KeyName)
```
I would like to see a call to `gaia0.GetBalance` to assert the funds were sent, in both tests here.

mark-rushakoff commented on chain/cosmos/chain_node.go:
```diff
@@ -321,6 +321,18 @@ func (tn *ChainNode) CreateKey(ctx context.Context, name string) error {
+// RecoverKey restores a key from a given mnemonic.
+func (tn *ChainNode) RecoverKey(ctx context.Context, keyName, mnemonic string) error {
+	command := []string{
+		tn.Chain.Config().GetShell(),
```
This can be hardcoded to `"sh"` after removing the `GetShell` method.

mark-rushakoff commented on chain/cosmos/chain_node.go:
```diff
@@ -321,6 +321,18 @@ func (tn *ChainNode) CreateKey(ctx context.Context, name string) error {
+	command := []string{
+		tn.Chain.Config().GetShell(),
+		"-c",
+		fmt.Sprintf(`echo "%s" | %s keys add %s --recover --keyring-backend %s --home %s --output json`, mnemonic, tn.Chain.Config().Bin, keyName, keyring.BackendTest, tn.NodeHome()),
```
`echo %q | ...` should be safer, as that will surround the string with quotes and add some escaping of special characters.

chatton replied:
Thanks for this, I wasn't familiar with %q here!

mark-rushakoff reviewed (commented):
There are a few things to clean up, but I think we'll be able to get this merged soon.

DavidNix reviewed (commented):
Apologies, I did not mean to approve previously, just comment.

chatton commented:
@mark-rushakoff ready for another look, I've addressed your feedback, thanks for taking a look so quickly! 

Tests seem to be failing but it looks like an environmental issue on the test host?

agouin commented:
Hi @chatton, looking at the test failure logs:

> tn.NodeJob undefined (type *ChainNode has no field or method NodeJob)

Earlier today, `NodeJob` was replaced by `Exec` in #173, so those calls will need to be updated in `chain_node.go`.

chatton commented:
@agouin thanks for letting me know! I had missed that PR being merged.
